base/boot.jl

```diff
@@ -274,10 +274,10 @@ struct BoundsError <: Exception
     a::Any
     i::Any
     BoundsError() = new()
-    # For now, always copy arrays to avoid escaping them
-    # Eventually, we want to figure out if the copy is needed to save the performance of copying
-    # (i.e., if a escapes elsewhere, don't bother to make a copy)
-
+    # maybecopy --- non-semantic copy
+    # if escape analysis proves that this throw is the only place where an object would escape local scope,
+    # creates a copy to avoid that escape and enable memory optimization through memory_opt!
+    # otherwise if there are other escapes, maybecopy does not copy and just passes the object
     BoundsError(@nospecialize(a)) = (@noinline;
         a isa Array ? new(Core.maybecopy(a)) : new(a))
```

Comment:
Can we extend this beyond array? This is very much a problem for `MArray` and would greatly improve it's usability on the GPU.

Reply:
If we can enforce `copy` implementation for anything implementing `getindex`.
But note that all the purpose of this `maybecopy` is that it allows EA to skip possible escapes through `arrayref` and thus this `maybecopy` implementation is very specific to `Array`s. This could also be generalized if we implement some specific interface which tells EA that user type object is always copied upon `BoundsError` though.

Having said that, I'm not really fan of changing the program semantics this way. Why not always making `copy` of it? `BoundsError` doesn't need to be high-performing.

Reply:
also making the default behavior of `maybecopy` no-op seems bad.. the default should be copy and we optimize it as no-op when the array is already escaped somewhere else.

Reply:
I do think the plan is eventually to have similar functionality as ImmutableArrays for other types, but yeah Shuhei is right that this initial PR is very Array-centric.

As for the non-semantic copy, I agree that the default should be copy - I misremembered that this was the approach we were taking (possibly because of some old incorrect test I had written). I’ll correct it shortly.

Reply:
No hurry from my side ;)
We can also talk with other core devs again to discuss if we can make copies by default?

Reply:
Update: https://github.com/JuliaLang/julia/pull/43738 is our new solution to this